File: georocket-server/src/test/java/io/georocket/storage/h2/H2StoreTest.java
package io.georocket.storage.h2;

import java.io.IOException;

import org.junit.After;
import org.junit.Before;
import org.junit.Rule;
import org.junit.rules.TemporaryFolder;

import io.georocket.constants.ConfigConstants;
import io.georocket.storage.StorageTest;
import io.georocket.util.PathUtils;
import io.vertx.core.AsyncResult;
import io.vertx.core.Future;
import io.vertx.core.Handler;
import io.vertx.core.Vertx;
import io.vertx.core.json.JsonObject;
import io.vertx.ext.unit.TestContext;

/**
 * Test {@link H2Store}
 * @author Michel Kraemer
 */
public class H2StoreTest extends StorageTest {
  /**
   * Create a temporary tempFolder
   */
  @Rule
  public TemporaryFolder tempFolder = new TemporaryFolder();

  private String path;
  private LegacyH2Store store;
  
  /**
   * Set up the test
   * @throws IOException if a temporary file could not be created
   */
  @Before
  public void setUp() throws IOException {
    path = tempFolder.newFile().getAbsolutePath();
  }
  
  /**
   * Release test resources
   */
  @After
  public void tearDown() {
    if (store != null) {
      store.close();
      store = null;
    }
  }
  
  private void configureVertx(Vertx vertx) {
    JsonObject config = vertx.getOrCreateContext().config();
    config.put(ConfigConstants.STORAGE_H2_PATH, path);
  }

  @Override
  protected LegacyH2Store createStore(Vertx vertx) {
    if (store == null) {
      configureVertx(vertx);
      store = new LegacyH2Store(vertx);
    }
    return store;
  }

  @Override
  protected void prepareData(TestContext context, Vertx vertx, String path, Handler<AsyncResult<String>> handler) {
    String p = PathUtils.join(path, ID);
    createStore(vertx).getMap().put(p, CHUNK_CONTENT);
    handler.handle(Future.succeededFuture(p));
  }

  @Override
  protected void validateAfterStoreAdd(TestContext context, Vertx vertx, String path,
    Handler<AsyncResult<Void>> handler) {
    context.assertEquals(1, createStore(vertx).getMap().size());
    handler.handle(Future.succeededFuture());
  }

  @Override
  protected void validateAfterStoreDelete(TestContext context, Vertx vertx, String path,
    Handler<AsyncResult<Void>> handler) {
    context.assertEquals(0, createStore(vertx).getMap().size());
    handler.handle(Future.succeededFuture());
  }
}

File: georocket-server/src/test/kotlin/io/georocket/storage/h2/H2StoreTest.kt
package io.georocket.storage.h2

import io.georocket.storage.StorageTest
import io.georocket.util.PathUtils
import io.vertx.core.Vertx
import io.vertx.junit5.VertxTestContext
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.AfterEach
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.io.TempDir
import java.io.File
import java.nio.file.Path

/**
 * Test [H2Store]
 * @author Michel Kraemer
 */
class H2StoreTest : StorageTest() {
  private lateinit var path: String
  private lateinit var store: H2Store

  /**
   * Set up the test
   */
  @BeforeEach
  fun setUp(vertx: Vertx, @TempDir tempFolder: Path) {
    path = File(tempFolder.toFile(), "h2").absolutePath
    store = H2Store(vertx, path)
  }

  /**
   * Release test resources
   */
  @AfterEach
  fun tearDown() {
    store.close()
  }

  override fun createStore(vertx: Vertx): H2Store {
    return store
  }

  override suspend fun prepareData(ctx: VertxTestContext, vertx: Vertx,
      path: String?): String {
    val p = PathUtils.join(path, ID)
    store.map[p] = CHUNK_CONTENT
    return p
  }

  override suspend fun validateAfterStoreAdd(ctx: VertxTestContext,
      vertx: Vertx, path: String?) {
    ctx.verify {
      assertThat(store.map).hasSize(1)
    }
  }

  override suspend fun validateAfterStoreDelete(ctx: VertxTestContext,
      vertx: Vertx, path: String) {
    ctx.verify {
      assertThat(store.map).hasSize(0)
    }
  }
}
